File: go/config/bootstrap/bootstrap_test.go
package bootstrap

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

type capturedReq struct {
	method string
	url    string
	header http.Header
	body   string
}

type recorder struct {
	t         *testing.T
	responses []func(w http.ResponseWriter, r *http.Request)
	calls     []capturedReq
}

func (r *recorder) handle(w http.ResponseWriter, req *http.Request) {
	body, _ := io.ReadAll(req.Body)
	r.calls = append(r.calls, capturedReq{
		method: req.Method,
		url:    req.URL.String(),
		header: req.Header.Clone(),
		body:   string(body),
	})
	if len(r.responses) == 0 {
		r.t.Fatalf("recorder ran out of queued responses")
	}
	fn := r.responses[0]
	r.responses = r.responses[1:]
	fn(w, req)
}

func ok(body string) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, body)
	}
}

func fail(status int, body string) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(status)
		_, _ = io.WriteString(w, body)
	}
}

func newEnvMap(pairs ...string) map[string]string {
	m := map[string]string{}
	for i := 0; i+1 < len(pairs); i += 2 {
		m[pairs[i]] = pairs[i+1]
	}
	return m
}

func envFn(m map[string]string) func(string) string {
	return func(k string) string { return m[k] }
}

func newRecorderServer(t *testing.T, responses ...func(w http.ResponseWriter, r *http.Request)) (*httptest.Server, *recorder) {
	rec := &recorder{t: t, responses: responses}
	srv := httptest.NewServer(http.HandlerFunc(rec.handle))
	return srv, rec
}

func baseEnv(serverURL string) map[string]string {
	return newEnvMap(
		"SMOOAI_CONFIG_API_URL", serverURL,
		"SMOOAI_CONFIG_AUTH_URL", serverURL,
		"SMOOAI_CONFIG_CLIENT_ID", "client-id-123",
		"SMOOAI_CONFIG_CLIENT_SECRET", "client-secret-456",
		"SMOOAI_CONFIG_ORG_ID", "org-789",
	)
}

func TestFetch_ReturnsValueForKnownKey(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"TOKEN"}`),
		ok(`{"values":{"databaseUrl":"postgres://x"}}`),
	)
	defer srv.Close()

	v, err := Fetch(context.Background(), "databaseUrl",
		withGetEnv(envFn(baseEnv(srv.URL))),
	)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if v != "postgres://x" {
		t.Fatalf("got %q, want postgres://x", v)
	}
	if len(rec.calls) != 2 {
		t.Fatalf("got %d HTTP calls, want 2", len(rec.calls))
	}
}

func TestFetch_ReturnsEmptyForMissingKey(t *testing.T) {
	resetCache()
	srv, _ := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		ok(`{"values":{"other":"x"}}`),
	)
	defer srv.Close()

	v, err := Fetch(context.Background(), "databaseUrl",
		withGetEnv(envFn(baseEnv(srv.URL))),
	)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if v != "" {
		t.Fatalf("got %q, want empty string", v)
	}
}

func TestFetch_CachesValuesPerEnv(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		ok(`{"values":{"a":"1","b":"2"}}`),
	)
	defer srv.Close()

	ctx := context.Background()
	env := envFn(baseEnv(srv.URL))

	v1, err := Fetch(ctx, "a", withGetEnv(env))
	if err != nil || v1 != "1" {
		t.Fatalf("a: %q err=%v", v1, err)
	}
	v2, err := Fetch(ctx, "b", withGetEnv(env))
	if err != nil || v2 != "2" {
		t.Fatalf("b: %q err=%v", v2, err)
	}
	if len(rec.calls) != 2 {
		t.Fatalf("expected 2 HTTP calls (cached), got %d", len(rec.calls))
	}
}

func TestFetch_RefetchesOnEnvChange(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"T1"}`),
		ok(`{"values":{"a":"dev"}}`),
		ok(`{"access_token":"T2"}`),
		ok(`{"values":{"a":"prod"}}`),
	)
	defer srv.Close()
	env := envFn(baseEnv(srv.URL))

	v1, err := Fetch(context.Background(), "a", withGetEnv(env), WithEnvironment("development"))
	if err != nil || v1 != "dev" {
		t.Fatalf("v1: %q err=%v", v1, err)
	}
	v2, err := Fetch(context.Background(), "a", withGetEnv(env), WithEnvironment("production"))
	if err != nil || v2 != "prod" {
		t.Fatalf("v2: %q err=%v", v2, err)
	}
	if len(rec.calls) != 4 {
		t.Fatalf("expected 4 HTTP calls, got %d", len(rec.calls))
	}
}

func TestFetch_OAuthRequestShape(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"TOKEN"}`),
		ok(`{"values":{"k":"v"}}`),
	)
	defer srv.Close()

	_, err := Fetch(context.Background(), "k", withGetEnv(envFn(baseEnv(srv.URL))))
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	auth := rec.calls[0]
	if auth.method != "POST" {
		t.Fatalf("auth method = %s, want POST", auth.method)
	}
	if auth.url != "/token" {
		t.Fatalf("auth url = %s, want /token", auth.url)
	}
	if !strings.Contains(auth.body, "grant_type=client_credentials") ||
		!strings.Contains(auth.body, "client_id=client-id-123") ||
		!strings.Contains(auth.body, "client_secret=client-secret-456") ||
		!strings.Contains(auth.body, "provider=client_credentials") {
		t.Fatalf("auth body missing required fields: %s", auth.body)
	}
}

func TestFetch_ValuesRequestShape(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"TOKEN"}`),
		ok(`{"values":{"k":"v"}}`),
	)
	defer srv.Close()

	_, err := Fetch(context.Background(), "k",
		withGetEnv(envFn(baseEnv(srv.URL))),
		WithEnvironment("staging env"),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	values := rec.calls[1]
	if values.method != "GET" {
		t.Fatalf("values method = %s, want GET", values.method)
	}
	wantPath := "/organizations/org-789/config/values?environment=staging+env"
	// QueryEscape encodes spaces as +; both + and %20 are valid. Accept either.
	if values.url != wantPath && values.url != "/organizations/org-789/config/values?environment=staging%20env" {
		t.Fatalf("values url = %s, want %s", values.url, wantPath)
	}
	if got := values.header.Get("Authorization"); got != "Bearer TOKEN" {
		t.Fatalf("Authorization = %q, want Bearer TOKEN", got)
	}
}

func TestFetch_MissingCredsErrors(t *testing.T) {
	resetCache()
	env := baseEnv("http://example.test")
	delete(env, "SMOOAI_CONFIG_CLIENT_ID")

	_, err := Fetch(context.Background(), "k", withGetEnv(envFn(env)))
	if err == nil || !strings.Contains(err.Error(), "CLIENT_ID,CLIENT_SECRET,ORG_ID") {
		t.Fatalf("got err=%v, want missing-creds error", err)
	}
}

func TestFetch_AcceptsLegacyApiKey(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		ok(`{"values":{"k":"v"}}`),
	)
	defer srv.Close()

	env := baseEnv(srv.URL)
	delete(env, "SMOOAI_CONFIG_CLIENT_SECRET")
	env["SMOOAI_CONFIG_API_KEY"] = "legacy-secret"

	v, err := Fetch(context.Background(), "k", withGetEnv(envFn(env)))
	if err != nil || v != "v" {
		t.Fatalf("v=%q err=%v", v, err)
	}
	if !strings.Contains(rec.calls[0].body, "client_secret=legacy-secret") {
		t.Fatalf("body did not include legacy client_secret: %s", rec.calls[0].body)
	}
}

func TestFetch_AcceptsLegacyAuthUrl(t *testing.T) {
	resetCache()
	srv, rec := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		ok(`{"values":{"k":"v"}}`),
	)
	defer srv.Close()

	env := baseEnv(srv.URL)
	delete(env, "SMOOAI_CONFIG_AUTH_URL")
	env["SMOOAI_AUTH_URL"] = srv.URL

	if _, err := Fetch(context.Background(), "k", withGetEnv(envFn(env))); err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if rec.calls[0].url != "/token" {
		t.Fatalf("legacy SMOOAI_AUTH_URL not used; first call URL = %s", rec.calls[0].url)
	}
}

func TestFetch_OAuthFailureErrors(t *testing.T) {
	resetCache()
	srv, _ := newRecorderServer(t, fail(401, "invalid_client"))
	defer srv.Close()

	_, err := Fetch(context.Background(), "k", withGetEnv(envFn(baseEnv(srv.URL))))
	if err == nil || !strings.Contains(err.Error(), "OAuth token exchange failed: HTTP 401") {
		t.Fatalf("got err=%v, want OAuth 401 error", err)
	}
}

func TestFetch_ValuesFailureErrors(t *testing.T) {
	resetCache()
	srv, _ := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		fail(500, "boom"),
	)
	defer srv.Close()

	_, err := Fetch(context.Background(), "k", withGetEnv(envFn(baseEnv(srv.URL))))
	if err == nil || !strings.Contains(err.Error(), "GET /config/values failed: HTTP 500") {
		t.Fatalf("got err=%v, want values 500 error", err)
	}
}

func TestFetch_OAuthMissingAccessToken(t *testing.T) {
	resetCache()
	srv, _ := newRecorderServer(t, ok(`{}`))
	defer srv.Close()

	_, err := Fetch(context.Background(), "k", withGetEnv(envFn(baseEnv(srv.URL))))
	if err == nil || !strings.Contains(err.Error(), "no access_token") {
		t.Fatalf("got err=%v, want no-access-token error", err)
	}
}

func TestResolveEnv_Variants(t *testing.T) {
	cases := []struct {
		name     string
		explicit string
		env      map[string]string
		want     string
	}{
		{"explicit wins", "explicit", newEnvMap("SST_STAGE", "ignored"), "explicit"},
		{"SST_STAGE", "", newEnvMap("SST_STAGE", "brentrager"), "brentrager"},
		{"NEXT_PUBLIC_SST_STAGE", "", newEnvMap("NEXT_PUBLIC_SST_STAGE", "dev-stage"), "dev-stage"},
		{"SST_RESOURCE_App", "", newEnvMap("SST_RESOURCE_App", `{"stage":"sst-resource-stage"}`), "sst-resource-stage"},
		{"production stays production", "", newEnvMap("SST_STAGE", "production"), "production"},
		{"SMOOAI_CONFIG_ENV fallback", "", newEnvMap("SMOOAI_CONFIG_ENV", "qa"), "qa"},
		{"development default", "", newEnvMap(), "development"},
		{"malformed SST_RESOURCE_App falls through", "", newEnvMap("SST_RESOURCE_App", "{not json", "SMOOAI_CONFIG_ENV", "qa"), "qa"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := resolveEnv(envFn(tc.env), tc.explicit)
			if got != tc.want {
				t.Fatalf("got %q, want %q", got, tc.want)
			}
		})
	}
}

func TestFetch_StringifiesNonStringValues(t *testing.T) {
	resetCache()
	srv, _ := newRecorderServer(t,
		ok(`{"access_token":"T"}`),
		ok(`{"values":{"count":42,"flag":true,"pi":3.5}}`),
	)
	defer srv.Close()
	env := envFn(baseEnv(srv.URL))

	c, _ := Fetch(context.Background(), "count", withGetEnv(env))
	if c != "42" {
		t.Fatalf("count = %q, want 42", c)
	}
	f, _ := Fetch(context.Background(), "flag", withGetEnv(env))
	if f != "true" {
		t.Fatalf("flag = %q, want true", f)
	}
	p, _ := Fetch(context.Background(), "pi", withGetEnv(env))
	if p != "3.5" {
		t.Fatalf("pi = %q, want 3.5", p)
	}
}

File: src/bootstrap/__tests__/bootstrap.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { resetBootstrapCacheForTests, bootstrapFetch } from '../index';

// Snapshot + restore process.env for full test isolation.
const originalEnv = { ...process.env };

function clearSmooEnv() {
    for (const key of Object.keys(process.env)) {
        if (key.startsWith('SMOOAI_') || key.startsWith('SST_') || key === 'NEXT_PUBLIC_SST_STAGE') {
            delete process.env[key];
        }
    }
}

function setBaseEnv() {
    process.env.SMOOAI_CONFIG_API_URL = 'https://api.example.test';
    process.env.SMOOAI_CONFIG_AUTH_URL = 'https://auth.example.test';
    process.env.SMOOAI_CONFIG_CLIENT_ID = 'client-id-123';
    process.env.SMOOAI_CONFIG_CLIENT_SECRET = 'client-secret-456';
    process.env.SMOOAI_CONFIG_ORG_ID = 'org-789';
}

type FetchArgs = [input: string | URL, init?: RequestInit];

function mockFetchResponses(responses: Array<{ ok?: boolean; status?: number; body: unknown; text?: string }>) {
    const fetchMock = vi.fn(async (..._args: FetchArgs) => {
        const next = responses.shift();
        if (!next) throw new Error('mock fetch ran out of queued responses');
        const ok = next.ok ?? true;
        const status = next.status ?? (ok ? 200 : 500);
        return {
            ok,
            status,
            json: async () => next.body,
            text: async () => next.text ?? JSON.stringify(next.body),
        } as unknown as Response;
    });
    vi.stubGlobal('fetch', fetchMock);
    return fetchMock;
}

describe('bootstrapFetch', () => {
    beforeEach(() => {
        clearSmooEnv();
        resetBootstrapCacheForTests();
        setBaseEnv();
    });

    afterEach(() => {
        vi.unstubAllGlobals();
        // restore env
        for (const key of Object.keys(process.env)) {
            if (!(key in originalEnv)) delete process.env[key];
        }
        for (const [k, v] of Object.entries(originalEnv)) {
            process.env[k] = v;
        }
    });

    it('returns the value for a known key', async () => {
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { databaseUrl: 'postgres://example' } } }]);

        const value = await bootstrapFetch('databaseUrl');
        expect(value).toBe('postgres://example');
        expect(fetchMock).toHaveBeenCalledTimes(2);
    });

    it('returns undefined for a missing key without throwing', async () => {
        mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { otherKey: 'x' } } }]);

        const value = await bootstrapFetch('databaseUrl');
        expect(value).toBeUndefined();
    });

    it('caches the values map per env across calls', async () => {
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { a: '1', b: '2' } } }]);

        const a = await bootstrapFetch('a');
        const b = await bootstrapFetch('b');
        expect(a).toBe('1');
        expect(b).toBe('2');
        // Only 2 HTTP calls total (token + values), not 4.
        expect(fetchMock).toHaveBeenCalledTimes(2);
    });

    it('re-fetches when the resolved env changes', async () => {
        const fetchMock = mockFetchResponses([
            { body: { access_token: 'T1' } },
            { body: { values: { a: 'dev' } } },
            { body: { access_token: 'T2' } },
            { body: { values: { a: 'prod' } } },
        ]);

        const a1 = await bootstrapFetch('a', { environment: 'development' });
        const a2 = await bootstrapFetch('a', { environment: 'production' });
        expect(a1).toBe('dev');
        expect(a2).toBe('prod');
        expect(fetchMock).toHaveBeenCalledTimes(4);
    });

    it('sends OAuth client_credentials with the expected body', async () => {
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { k: 'v' } } }]);
        await bootstrapFetch('k');

        const [authUrl, authInit] = fetchMock.mock.calls[0]!;
        expect(authUrl).toBe('https://auth.example.test/token');
        expect(authInit!.method).toBe('POST');
        const body = String(authInit!.body);
        expect(body).toContain('grant_type=client_credentials');
        expect(body).toContain('client_id=client-id-123');
        expect(body).toContain('client_secret=client-secret-456');
        expect(body).toContain('provider=client_credentials');
    });

    it('sends the values GET with bearer token and url-encoded environment', async () => {
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { k: 'v' } } }]);
        await bootstrapFetch('k', { environment: 'staging env' });

        const [valuesUrl, valuesInit] = fetchMock.mock.calls[1]!;
        expect(valuesUrl).toBe('https://api.example.test/organizations/org-789/config/values?environment=staging%20env');
        const headers = valuesInit!.headers as Record<string, string>;
        expect(headers.Authorization).toBe('Bearer TOKEN');
    });

    it('throws when required env vars are missing', async () => {
        delete process.env.SMOOAI_CONFIG_CLIENT_ID;
        await expect(bootstrapFetch('k')).rejects.toThrow(/SMOOAI_CONFIG_\{CLIENT_ID,CLIENT_SECRET,ORG_ID\}/);
    });

    it('accepts legacy SMOOAI_CONFIG_API_KEY as client secret', async () => {
        delete process.env.SMOOAI_CONFIG_CLIENT_SECRET;
        process.env.SMOOAI_CONFIG_API_KEY = 'legacy-secret';
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { k: 'v' } } }]);
        const value = await bootstrapFetch('k');
        expect(value).toBe('v');
        const body = String(fetchMock.mock.calls[0]![1]!.body);
        expect(body).toContain('client_secret=legacy-secret');
    });

    it('accepts legacy SMOOAI_AUTH_URL when SMOOAI_CONFIG_AUTH_URL absent', async () => {
        delete process.env.SMOOAI_CONFIG_AUTH_URL;
        process.env.SMOOAI_AUTH_URL = 'https://legacy-auth.example.test';
        const fetchMock = mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { body: { values: { k: 'v' } } }]);
        await bootstrapFetch('k');
        expect(fetchMock.mock.calls[0]![0]).toBe('https://legacy-auth.example.test/token');
    });

    it('throws a clear error when the OAuth call fails', async () => {
        mockFetchResponses([{ ok: false, status: 401, body: {}, text: 'invalid_client' }]);
        await expect(bootstrapFetch('k')).rejects.toThrow(/OAuth token exchange failed: HTTP 401 invalid_client/);
    });

    it('throws a clear error when the values call fails', async () => {
        mockFetchResponses([{ body: { access_token: 'TOKEN' } }, { ok: false, status: 500, body: {}, text: 'boom' }]);
        await expect(bootstrapFetch('k')).rejects.toThrow(/GET \/config\/values failed: HTTP 500 boom/);
    });

    it('throws when the OAuth response is missing access_token', async () => {
        mockFetchResponses([{ body: {} }]);
        await expect(bootstrapFetch('k')).rejects.toThrow(/no access_token/);
    });

    describe('environment resolution', () => {
        it('uses explicit options.environment first', async () => {
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            process.env.SST_STAGE = 'should-be-ignored';
            await bootstrapFetch('k', { environment: 'explicit-env' });
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=explicit-env');
        });

        it('uses SST_STAGE when no explicit env', async () => {
            process.env.SST_STAGE = 'brentrager';
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=brentrager');
        });

        it('uses NEXT_PUBLIC_SST_STAGE as a fallback for stage', async () => {
            process.env.NEXT_PUBLIC_SST_STAGE = 'dev-stage';
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=dev-stage');
        });

        it('parses SST_RESOURCE_App JSON for stage', async () => {
            process.env.SST_RESOURCE_App = JSON.stringify({ stage: 'sst-resource-stage' });
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=sst-resource-stage');
        });

        it('maps stage=production to environment=production', async () => {
            process.env.SST_STAGE = 'production';
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=production');
        });

        it('uses SMOOAI_CONFIG_ENV when no stage env vars are set', async () => {
            process.env.SMOOAI_CONFIG_ENV = 'qa';
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=qa');
        });

        it('falls back to development as the last resort', async () => {
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=development');
        });

        it('survives malformed SST_RESOURCE_App JSON', async () => {
            process.env.SST_RESOURCE_App = '{not json';
            process.env.SMOOAI_CONFIG_ENV = 'qa';
            const fetchMock = mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: {} } }]);
            await bootstrapFetch('k');
            expect(fetchMock.mock.calls[1]![0]).toContain('environment=qa');
        });
    });

    it('stringifies non-string values', async () => {
        mockFetchResponses([{ body: { access_token: 'T' } }, { body: { values: { count: 42, flag: true } } }]);
        expect(await bootstrapFetch('count')).toBe('42');
        // cache hit, no new mocks needed
        expect(await bootstrapFetch('flag')).toBe('true');
    });
});
